x-pack/plugin/autoscaling/src/main/java/org/elasticsearch/xpack/autoscaling/storage/ReactiveStorageDeciderService.java

```diff
@@ -301,28 +310,77 @@ && canAllocate(shard, allocation) == false) {
             }
 
             // track these to ensure we do not double account if they both cannot remain and allocated due to storage.
-            Set<ShardRouting> unmovableShards = candidates.stream()
+            List<ShardNodeDecisions> unmovableShardNodeDecisions = candidates.stream()
                 .filter(s -> allocatedToTier(s, allocation))
-                .filter(s -> cannotRemainDueToStorage(s, allocation))
-                .collect(Collectors.toSet());
+                .flatMap(shard -> cannotRemainDueToStorage(shard, allocation).stream())
+                .toList();
+            List<ShardNodeDecisions> otherNodesOnTierAllocationDecisions = unmovableShardNodeDecisions.stream()
+                .map(ShardNodeDecisions::shard)
+                .map(
+                    shard -> new ShardNodeDecisions(
+                        shard,
+                        state.getRoutingNodes()
+                            .stream()
+                            .filter(n -> nodeTierPredicate.test(n.node()))
+                            .filter(n -> shard.currentNodeId().equals(n.nodeId()) == false)
+                            .map(n -> new NodeDecision(n.node(), allocationDeciders.canAllocate(shard, n, allocation)))
```

Comment:
Can we apply the limit when collecting these, both in order to limit memory use but also not calculate these decisions unnecesarily? We might need to collect both the cannot remain reason and the can allocate decision here then but I think that is fine?